KAFKA-4814: Enable ZK ACLs only when zookeeper.set.acl is set

rajinisivaram opened the pull request.

rajinisivaram committed:
KAFKA-4814: Enable ZK ACLs only when zookeeper.set.acl is set

ijuma reviewed (commented):
Thanks for the PR. Would it be possible to update some of the existing tests to verify this change?

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -109,7 +109,7 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
                       sessionTimeout = zkSessionTimeOutMs,
                       connectionTimeout = zkConnectionTimeoutMs,
-                      JaasUtils.isZkSecurityEnabled())
+                      kafkaConfig.zkEnableSecureAcls)
```
We need to check that this doesn't cause issues when `SimpleAclAuthorizer` is used by `AclCommand`. We may have to set this config in `AclCommand`.

rajinisivaram replied:
Have updated default properties.

rajinisivaram committed:
KAFKA-4814: Address review comments

rajinisivaram commented:
@ijuma Thank you for the review. I have added a check to two existing tests, one for secure acls and another for insecure. There was an issue with `controller_epoch` - it was being created in `KafkaController` without any ACLs even though it is listed as a secure path in `ZkUtils`. I think it was an oversight and fixed it, but it will be good if you can check, just in case it was deliberate. Since the migrator tool does add ACL to it, the change shouldn't cause any issues.

ijuma commented on core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala:
```diff
@@ -27,9 +28,15 @@ class SaslPlainPlaintextConsumerTest extends BaseConsumerTest with SaslTestHarne
+  @Before
+  override def setUp {
+    super.setUp
+    TestUtils.verifyUnsecureZkAcls(zkUtils)
```
It's a bit odd to have this check in `setUp`. Since `setUp` is always the same, we can probably just add a single test that verifies this. If we break the code, we would have one test failure with a clear cause. Same for the other test class.

rajinisivaram replied:
Done for both.

ijuma commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1135,6 +1136,71 @@ object TestUtils extends Logging {
+    info(s"ACL $acl")
+    acl.getPerms match {
+      case 1 => !sensitive && acl.getId.getScheme.equals("world")
+      case 31 => acl.getId.getScheme.equals("sasl")
```
Nit: `equals` -> `==`. There are a few other cases of that in this file.

rajinisivaram replied:
Done.

ijuma commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1135,6 +1136,71 @@ object TestUtils extends Logging {
+   * ZooKeeper code base.
+   */
+  def isAclSecure(acl: ACL, sensitive: Boolean): Boolean = {
+    info(s"ACL $acl")
```
I know this was already like this before, but maybe this and the other case should be `debug`?

rajinisivaram replied:
Done.

ijuma commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1135,6 +1136,71 @@ object TestUtils extends Logging {
+    secureZkPaths(zkUtils).foreach(path => {
+      if (zkUtils.pathExists(path)) {
+        val sensitive = ZkUtils.sensitivePath(path)
+        val aclCount = if (sensitive) usersWithAccess else usersWithAccess + 1
```
Maybe worth a comment explaining why we do this.

rajinisivaram replied:
Done.

ijuma commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1135,6 +1136,71 @@ object TestUtils extends Logging {
+        Seq.empty
+    }
+    val topLevelPaths = ZkUtils.SecureZkRootPaths ++ ZkUtils.SensitiveZkRootPaths
+    topLevelPaths.map(subPaths).flatten
```
Nit: `flatMap` instead of `map` followed by `flatten`.

rajinisivaram replied:
Done.

ijuma commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1135,6 +1136,71 @@ object TestUtils extends Logging {
+  private def secureZkPaths(zkUtils: ZkUtils) = {
+    def subPaths(path: String) = {
+      if (zkUtils.pathExists(path))
+        zkUtils.getChildren(path).map(c => path + "/" + c)
```
Should we be checking all children recursively?

rajinisivaram replied:
Thought there was no harm in checking all the paths.

ijuma replied:
Sorry, my question was whether we should check all the paths instead of stopping at the direct children. In other words, call `subPaths` for each child.

rajinisivaram replied:
Oops, sorry. I was talking about what I had intended to do rather than what I had actually done! Updated to call `subPaths` recursively. I had to update the `zkUtils` used for creating topics in the test which disables acls since the topics are also being tested.

ijuma commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1135,6 +1136,71 @@ object TestUtils extends Logging {
   /**
+   * Verifies that this ACL is the secure one. The
+   * values are based on the constants used in the
+   * ZooKeeper code base.
```
It would be nice if we linked to these constants in the ZK code base or at least explained them. Another existing issue, but it's quite difficult to understand what these magic values mean. :)

rajinisivaram replied:
Have switched to using ZK constants.

ijuma replied:
Much better, maybe we can remove the comment "The values are based on constants.." since it seems redundant.

rajinisivaram replied:
Done.

ijuma commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1135,6 +1136,71 @@ object TestUtils extends Logging {
+  }
+
+  /**
+   * Verifies that the ACL corresponds to the unsecure one.
```
It would be good to expand this a bit more. It seems that the definition if `unsecure` is a bit specific.

rajinisivaram replied:
Done.

ijuma reviewed (commented):
Thanks for the updates Rajini. I left a few minor comments to hopefully make the testing code a little easier to understand (most of them related to existing issues).

It may be worth running some security system tests to make sure things are still good after these changes.

rajinisivaram committed:
KAFKA-4814: Address review comments

rajinisivaram commented:
@ijuma Thank you for the review. I have made the updates.
I ran a few system tests locally and they were ok, I have scheduled a run of all the core tests (https://jenkins.confluent.io/job/system-test-kafka-branch-builder/), but waiting in a queue at the moment.

rajinisivaram commented:
Jenkins system test runs look broken at the moment (https://jenkins.confluent.io/job/system-test-kafka-branch-builder/840/). I ran `sanity_checks`, `zookeeper_security_upgrade_test` and `security_rolling_upgrade_test` locally and they passed. That covers a combination of with-acl and no-acl.

ijuma commented:
retest this please

ijuma commented on core/src/test/scala/integration/kafka/api/SaslPlainPlaintextConsumerTest.scala:
```diff
@@ -27,9 +28,23 @@ class SaslPlainPlaintextConsumerTest extends BaseConsumerTest with SaslTestHarne
+
+  @Before
+  override def setUp {
+    super.setUp
```
Nit: we can remove this.

rajinisivaram replied:
@ijuma Thank you for the reviews. I have made the changes.

ijuma reviewed (commented):
Thanks @rajinisivaram, a few trivial comments and we can then merge.

rajinisivaram committed:
KAFKA-4814: Address review comments

ijuma commented:
retest this please

ijuma reviewed (approved):
Thanks, LGTM.
